### PR TITLE
Feat/305 notification repo firestore

### LINKED ISSUE #305: Notification repository Firestore
Firestore implementation of the Notifications repository, following the established and merged interface (#292).

### PR DESCRIPTION
**Pull request by AACioffi**
## What?
Implements the production Firestore repository for notifications with comprehensive integration tests and security rules. Closes #305.

## Why?
The notification system requires:
- **Cloud persistence for cross-device/session access** - Notifications must persist in Firestore, not just in-memory
- **Security enforcement at the database level** - Users can only manage their own notifications while authenticated users can send notifications to others
- **Integration testing against real Firestore** - Verify actual database interactions using Firebase Emulator

## How?
**Implementation (`NotificationsRepositoryFirestore.kt`):**

- Implemented all `NotificationsRepository` interface methods using Firestore backend:
  - **Data structure**: Subcollections pattern `users/{userId}/notifications/{notificationId}` (scopes data by user)
  - `currentUserId()`: Wraps `Firebase.auth.currentUser?.uid`, throws `IllegalStateException` if unauthenticated (used in security checks)
  - `getNewId()`: ID generation via `collection(...).document().id`
  - `addNotification()`: Validates `recipientId` non-empty, converts object to Map, persists via `.set()`
  - `getUserNotifications()`: Queries user's subcollection with `.orderBy("emissionTime", DESCENDING)`, enforces caller matches requested userId (throws `SecurityException` if not)
  - `markAsRead()`: Partial field update `.update("wasRead", true)`
  - `getNotification()`: Retrieves by ID from current user's subcollection, throws `NoSuchElementException` if not found
  - `deleteNotification()`: Pre-fetches via `getNotification()` (verifies existence/ownership), then executes `.delete()`
- **Mappers**: Private helpers (snapshotToNotification and notificationToMap) that transform full Notification objects into Firestore-compatible Map<String, Any?> structures (and vice versa), handling nested type conversions like Timestamp <-> Date and Enum serialization.

**Security Rules (`firestore.rules`):**

- Added rule block for `/users/{userId}/notifications/{notificationId}`:
  - **Read**: Restricted to recipient only (`request.auth.uid == userId`)
  - **Create**: Any authenticated user can create (enables friend requests), but `recipientId` in data must match `userId` in path
  - **Update/Delete**: Restricted to recipient only (prevents senders from modifying/deleting after sending)

**Test Infrastructure (`FirestoreNotificationsGatherlyTest.kt`):**

- Base class for Notifications integration tests:
  - Configures connection to Firebase Emulator (Firestore + Auth)
  - Simulates multi-user scenarios with fake JWT tokens (Alice, Bob, Geoff, Donna)
  - Provides template notifications (`friendRequest1_template`, etc.) with configurable user IDs
  - Handles emulator cleanup and authentication state management
  - Signs in as user1 by default for simplified test setup

**Test Suite (`NotificationRepositoryFirestoreTest.kt`):**

- 17 comprehensive tests organized by category:
  - **Happy paths** (5 tests): ID uniqueness, CRUD operations, multi-user retrieval, mark as read, delete
  - **Security tests** (4 tests): Cross-user access attempts for `getUserNotifications()`, `getNotification()`, `deleteNotification()`, `markAsRead()` (all throw appropriate exceptions)
  - **Edge cases** (5 tests): Invalid IDs, empty lists, empty `recipientId` validation
  - **State verification & ordering** (3 tests): Field preservation, isolated updates (markAsRead only changes `wasRead`), descending timestamp sort

**Patterns:**

- Follows existing Firestore repository patterns
- Uses dynamic ID generation in tests (no hardcoded IDs)

## Testing?
All 17 tests pass against Firebase Emulator. Coverage metrics pending verification via CI.

## Screenshots

image

All tests pass.

image

Line coverage requirements fulfilled.

### CODE REVIEW

File: app/src/main/java/com/android/gatherly/model/notification/NotificationsRepositoryFirestore.kt [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+        .document(currentUserId())
+        .collection("notifications")
+        .document(notificationId)
+        .delete()
+        .await()
+  }
+
+  override suspend fun addNotification(notification: Notification) {
+    if (notification.recipientId.isEmpty()) {
+      throw IllegalArgumentException("Notification must have a valid recipientId")
```
Comment (claumnl) [inline app/src/main/java/com/android/gatherly/model/notification/NotificationsRepositoryFirestore.kt]: **NitPik:** Maybe consider making this comment "Notification can't have an empty recipientId" to make it easier to pinpoint the issue while debugging.

File: app/src/androidTest/java/com/android/gatherly/model/notification/NotificationsRepositoryFirestoreTest.kt [hunk 2]
```diff
@@ -0,0 +1,292 @@
...
+        eventReminder1_template.copy(id = repository.getNewId(), recipientId = user2Id)
+    repository.addNotification(notification)
+
+    // Try to get user2's notifications while signed in as user1
+    signInWithToken(user1Token)
+    repository.getUserNotifications(user2Id) // should throw SecurityException
+  }
+
+  @Test(expected = NoSuchElementException::class)
+  fun getNotification_forDifferentUser_throwsSecurityException() = runTest {
```
[the remark this answers is not in the document]
Comment (AACioffi) [reply app/src/androidTest/java/com/android/gatherly/model/notification/NotificationsRepositoryFirestoreTest.kt]: I will not. This is not a state verification test (of which there are 2 already), and I will therefore not check whether the fields were preserved.

File: app/src/main/java/com/android/gatherly/model/notification/NotificationsRepositoryFirestore.kt [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+   * signed in.
+   */
+  private fun currentUserId(): String {
+    return Firebase.auth.currentUser?.uid ?: throw IllegalStateException("No signed in user")
+  }
+
+  /** Returns a new unique ID for a notification. */
+  override fun getNewId(): String {
+    return db.collection("users")
+        .document(currentUserId())
```
[the remark this answers is not in the document]
Comment (AACioffi) [reply app/src/main/java/com/android/gatherly/model/notification/NotificationsRepositoryFirestore.kt]: Good observation. You're right about the mechanics, but this doesn't cause issues in practice:

- `getNewId()` just generates a random unique string via Firestore's `.document().id` - it doesn't create a document or check the collection
- `addNotification()` correctly stores the notification in the recipient's collection using `notification.recipientId` (line 62)
- The ViewModel is responsible for creating the Notification object with the correct recipientId

So in your Alice and Bob example, even though the ID is generated from Alice's path, the notification ends up correctly in Bob's collection because the Notification object has recipientId `Bob's ID`.

[the remark this answers is not in the document]
Comment (claumnl) [reply app/src/main/java/com/android/gatherly/model/notification/NotificationsRepositoryFirestore.kt]: All good then!
